File: leetcode_python/symmetric_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {boolean}
    def isSymmetric(self, root):
        if not root:
            return True
        return self.isMirror(root.left, root.right)

    def isMirror(self, node1, node2):
        if not node1 and not node2: # both are null
            return True
        elif bool(node1) ^ bool(node2): # one is null
            return False
        else: # both not null
            return node1.val == node2.val and self.isMirror(node1.left, node2.right) and self.isMirror(node1.right, node2.left)
        

File: leetcode_cpp/symmetric_tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isMirror(TreeNode *left, TreeNode *right) {
        if (!left and !right) return true;
        else if (!left) return false;
        else if (!right) return false;

        if (left -> val != right -> val) {
            return false;
        }
        else if (!isMirror(left -> left, right -> right)) {
            return false;
        }
        else return isMirror(left -> right, right -> left);
    }

    bool isSymmetric(TreeNode *root) {
        if (!root) return true;
        return isMirror(root -> left, root -> right);
    }
};
